File: flutter/ios/traccar_client_sdk/Sources/traccar_client_sdk/TraccarClientSdkPlugin.swift
import Flutter
import UIKit
import TraccarClientSDK

public class TraccarClientSdkPlugin: NSObject, FlutterPlugin {
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "traccar_client_sdk", binaryMessenger: registrar.messenger())
    let instance = TraccarClientSdkPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "start":
      let args = call.arguments as! [String: Any]
      let config = parseConfig(args)
      Task {
        let tracker = try await TrackerKt.sharedTracker()
        try await tracker.start(config: config)
        result(true)
      }
    case "stop":
      Task {
        let tracker = try await TrackerKt.sharedTracker()
        try await tracker.stop()
        result(nil)
      }
    case "requestPosition":
      let args = call.arguments as! [String: Any]
      let config = parseConfig(args)
      Task {
        let tracker = try await TrackerKt.sharedTracker()
        tracker.requestPosition(config: config)
        result(nil)
      }
    case "isTracking":
      Task {
        let tracker = try await TrackerKt.sharedTracker()
        result(tracker.isTracking.value)
      }
    case "getLogs":
      Task {
        let tracker = try await TrackerKt.sharedTracker()
        let entries = tracker.getLogs().map { ["time": $0.time, "message": $0.message] as [String: Any] }
        result(entries)
      }
    case "clearLogs":
      Task {
        let tracker = try await TrackerKt.sharedTracker()
        tracker.clearLogs()
        result(nil)
      }
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  private func parseConfig(_ args: [String: Any]) -> Config {
    let location = args["location"] as! [String: Any]
    let notification = args["notification"] as! [String: Any]
    return Config(
      serverUrl: args["serverUrl"] as! String,
      deviceId: args["deviceId"] as! String,
      location: LocationConfig(
        accuracy: parseAccuracy(location["accuracy"] as! String),
        distanceMeters: Int32(location["distanceMeters"] as! Int),
        intervalSeconds: Int32(location["intervalSeconds"] as! Int),
        angleDegrees: Int32(location["angleDegrees"] as! Int),
        stopDetection: location["stopDetection"] as! Bool,
        stopTimeoutSeconds: Int32(location["stopTimeoutSeconds"] as! Int),
        stationaryRadiusMeters: Int32(location["stationaryRadiusMeters"] as! Int)
      ),
      wakeLock: args["wakeLock"] as! Bool,
      buffer: args["buffer"] as! Bool,
      notification: NotificationConfig(text: notification["text"] as! String)
    )
  }

  private func parseAccuracy(_ name: String) -> Accuracy {
    switch name {
    case "HIGHEST": return Accuracy.highest
    case "HIGH": return Accuracy.high
    case "LOW": return Accuracy.low
    default: return Accuracy.medium
    }
  }
}

File: flutter/android/src/main/kotlin/org/traccar/traccar_client_sdk/TraccarClientSdkPlugin.kt
package org.traccar.traccar_client_sdk

import android.content.Context
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch
import org.traccar.client.Accuracy
import org.traccar.client.Config
import org.traccar.client.LocationConfig
import org.traccar.client.NotificationConfig
import org.traccar.client.sharedTracker
import org.traccar.client.startTracking

class TraccarClientSdkPlugin :
    FlutterPlugin,
    MethodCallHandler {

    private lateinit var channel: MethodChannel
    private lateinit var context: Context
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main)

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        context = binding.applicationContext
        channel = MethodChannel(binding.binaryMessenger, "traccar_client_sdk")
        channel.setMethodCallHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "start" -> {
                val config = parseConfig(call.arguments as Map<*, *>)
                scope.launch {
                    result.success(sharedTracker().startTracking(context, config))
                }
            }
            "stop" -> {
                scope.launch {
                    sharedTracker().stop()
                    result.success(null)
                }
            }
            "requestPosition" -> {
                val config = parseConfig(call.arguments as Map<*, *>)
                scope.launch {
                    sharedTracker().requestPosition(config)
                    result.success(null)
                }
            }
            "isTracking" -> {
                scope.launch {
                    result.success(sharedTracker().isTracking.value)
                }
            }
            "getLogs" -> {
                scope.launch {
                    val entries = sharedTracker().getLogs().map {
                        mapOf("time" to it.time, "message" to it.message)
                    }
                    result.success(entries)
                }
            }
            "clearLogs" -> {
                scope.launch {
                    sharedTracker().clearLogs()
                    result.success(null)
                }
            }
            else -> result.notImplemented()
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        scope.cancel()
    }

    private fun parseConfig(args: Map<*, *>): Config {
        val location = args["location"] as Map<*, *>
        val notification = args["notification"] as Map<*, *>
        return Config(
            serverUrl = args["serverUrl"] as String,
            deviceId = args["deviceId"] as String,
            location = LocationConfig(
                accuracy = Accuracy.valueOf(location["accuracy"] as String),
                distanceMeters = (location["distanceMeters"] as Number).toInt(),
                intervalSeconds = (location["intervalSeconds"] as Number).toInt(),
                angleDegrees = (location["angleDegrees"] as Number).toInt(),
                stopDetection = location["stopDetection"] as Boolean,
                stopTimeoutSeconds = (location["stopTimeoutSeconds"] as Number).toInt(),
                stationaryRadiusMeters = (location["stationaryRadiusMeters"] as Number).toInt(),
            ),
            wakeLock = args["wakeLock"] as Boolean,
            buffer = args["buffer"] as Boolean,
            notification = NotificationConfig(text = notification["text"] as String),
        )
    }
}
